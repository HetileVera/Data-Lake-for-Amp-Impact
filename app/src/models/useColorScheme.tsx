import { useTheme } from "@chakra-ui/react";

export function useColorScheme() {
  const colorScheme = useTheme().name;
  const tone = (value: number | undefined) => (value ? `${colorScheme}.${value}` : colorScheme);

  return { tone, colorScheme };
}
