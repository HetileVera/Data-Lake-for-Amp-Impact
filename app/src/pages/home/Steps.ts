export type TStepInfo = {
  title: string;
  desc: string;
};

export const steps: TStepInfo[] = [
  {
    title: "Connect to your AWS account",
    desc: `We need your AWS admin credentials so that we can connect to your AWS account and provision the resources for the data lake. We will
  guide you through the process of obtaining the needed credentials`,
  },
  {
    title: "Connect to your Amp Impact Salesforce organization",
    desc: `For AWS to access your Amp Impact Salesforce data, we will need your authorization. We will guide you through the steps of setting up Amazon AppFlow which
    is a service that allows AWS to connect to Salesforce, you will be creating a Salesforce connection, signing in to Salesforce and authorizing access`,
  },
  {
    title: "Select data import options",
    desc: `In this step, you can pick the standard Amp Impact import options for your Salesforce data or customize these options to fit your needs`,
  },
  {
    title: "Review and confirm",
    desc: `You get a chance to review and confirm the settings you have provided`,
  },
  {
    title: "Sit back and relax",
    desc: `In this step, we will do all the necessary provisioning and data importing. You can sit back and relax and watch the progress of the data import`,
  },
  {
    title: "Connect to Analytics",
    desc: "This is the final step, you can create users with credentials to access the data lake using Tableau",
  },
];
