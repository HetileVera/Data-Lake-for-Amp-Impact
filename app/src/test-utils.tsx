import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
