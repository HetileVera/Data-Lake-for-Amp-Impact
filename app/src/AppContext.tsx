import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
