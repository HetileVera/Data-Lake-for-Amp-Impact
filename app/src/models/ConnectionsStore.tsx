import { useEffect } from "react";
import isEmpty from "lodash/isEmpty";
import { types, Instance, getParent } from "mobx-state-tree";

import { useInstallation } from "AppContext";
import { listAppflowConnections } from "api/list-appflow-connections";

import { BaseStore, isStoreNew, isStoreError, isStoreReady, isStoreLoading, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

/**
 * Represents the store for the AppFlow Salesforce connection names
 */
export const ConnectionsStore = BaseStore.named("ConnectionsStore")
  .props({
    connectionNames: types.array(types.string),
  })

  .views((self) => ({
    get empty() {
      return isEmpty(self.connectionNames);
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },
  }))

  .actions((self) => {
    const superReset = self.reset;
    return {
      reset() {
        superReset();
        self.connectionNames.clear();
      },

      async doLoad() {
        const accessKey = self.credentials.accessKey;
        const secretKey = self.credentials.secretKey;
        const region = self.region;

        const rawConnectionNames = await listAppflowConnections({ accessKey, secretKey, region });
        self.runInAction(() => {
          self.connectionNames.replace(rawConnectionNames);
        });
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IConnectionsStore extends Instance<typeof ConnectionsStore> {}

export function useConnectionsStore() {
  const installation = useInstallation();
  const store = installation.connectionsStore;

  useEffect(() => {
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
