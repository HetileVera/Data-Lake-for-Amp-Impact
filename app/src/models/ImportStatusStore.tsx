import includes from "lodash/includes";
import split from "lodash/split";
import last from "lodash/last";
import chunkIt from "lodash/chunk";
import first from "lodash/first";
import isEmpty from "lodash/isEmpty";
import sortBy from "lodash/sortBy";
import { ExecutionStatus, HistoryEvent } from "@aws-sdk/client-sfn";
import { types, Instance, getParent } from "mobx-state-tree";
import { values } from "mobx";
import { useEffect } from "react";

import { useInstallation } from "AppContext";
import { s3GetJson } from "api/s3-get-json";
import { describeSFN } from "api/describe-sfn";
import { getExecutionFailureInfo, getSFNHistory, isExecutionFailureEvent } from "api/get-sfn-history";
import { s3ListObjects, S3Object } from "api/list-s3-objects";
import { getLatestSFN } from "api/get-lastest-sfn";
import { delay, niceByte, niceNumber } from "helpers/utils";

import { BaseStore, isStoreError, isStoreLoading, isStoreNew, isStoreReady, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

const isSuccessStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a success
  const successStatuses = ["SUCCEEDED"];
  return includes(successStatuses, status);
};

const isFailureStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a failure
  const failureStatuses = ["FAILED", "ABORTED", "TIMED_OUT"];
  return includes(failureStatuses, status);
};

export enum ObjectImportStatusCode {
  Not_Started = "NOT_STARTED",
  Preparing = "PREPARING",
  In_Progress = "IN_PROGRESS",
  Importing = "IMPORTING",
  Successful = "SUCCESSFUL",
}

interface ObjectStatusInput {
  name: string;
  label?: string;
  status?: ObjectImportStatusCode;
  rowCount?: number;
  columnCount?: number;
  importSeconds?: number;
  bytesWritten?: number;
}

interface RemoteObject {
  status: ObjectImportStatusCode;
  metadata?: {
    rowCount: number;
    columnCount: number;
    importSeconds: number;
    bytesWritten: number;
  };
}

/**
 * Represents a Salesforce object import status
 */
const ObjectImportStatus = types
  .model("ObjectImportStatus", {
    name: "",
    label: "",
    status: types.optional(
      types.enumeration<ObjectImportStatusCode>("ObjectImportStatusCode", Object.values(ObjectImportStatusCode)),
      ObjectImportStatusCode.Not_Started
    ),
    rowCount: 0,
    columnCount: 0,
    importSeconds: 0,
    bytesWritten: 0,
    stale: false,
  })

  .actions((self) => ({
    setObjectStatus({ name, label, status, rowCount, columnCount, importSeconds, bytesWritten }: ObjectStatusInput) {
      self.name = name;
      if (label) self.label = label;
      if (status) self.status = status;
      if (rowCount) self.rowCount = rowCount;
      if (columnCount) self.columnCount = columnCount;
      if (importSeconds) self.importSeconds = importSeconds;
      if (bytesWritten) self.bytesWritten = bytesWritten;
    },

    setStale(flag: boolean) {
      self.stale = flag;
      if (flag) self.status = ObjectImportStatusCode.Not_Started;
    },
  }))

  .views((self) => ({
    get isNotStarted(): boolean {
      return self.status === ObjectImportStatusCode.Not_Started;
    },

    get isPreparing(): boolean {
      return self.status === ObjectImportStatusCode.Preparing;
    },

    get isInProgress(): boolean {
      return self.status === ObjectImportStatusCode.In_Progress;
    },

    get isImporting(): boolean {
      return self.status === ObjectImportStatusCode.Importing;
    },

    get isSuccessful(): boolean {
      return self.status === ObjectImportStatusCode.Successful;
    },

    get warning(): boolean {
      return self.rowCount < 0;
    },
  }))

  .views((self) => ({
    get statsAvailable(): boolean {
      // Tells us if any of the stats such as (rowCount, et) are available.
      return self.isSuccessful;
    },
  }))

  .views((self) => ({
    get rowCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.rowCount);
    },

    get columnCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.columnCount);
    },

    get importSecondsDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return `${Math.round(self.importSeconds)} sec`;
    },

    get bytesWrittenDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceByte(self.bytesWritten);
    },
  }));

export interface IObjectImportStatus extends Instance<typeof ObjectImportStatus> {}

export enum LoadingStage {
  GettingSchemas = "GETTING_SCHEMAS",
  MonitoringProgress = "MONITORING_PROGRESS",
}

/**
 * Represents the model for the status of importing salesforce objects. This includes each object import status.
 */
export const ImportStatusStore = BaseStore.named("ImportStatusStore")
  .props({
    objects: types.map(ObjectImportStatus),
    startTime: 0, // The start time of the execution (epoch milliseconds)
    loadingStage: types.maybe(types.enumeration<LoadingStage>("LoadingStage", Object.values(LoadingStage))),
    _errorDetail: "", // Expected to be in markdown https://commonmark.org/help/
  })

  .views((self) => ({
    get empty() {
      return self.objects.size === 0;
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },

    get importWorkflowArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get metadataBucket(): string {
      const parent: IInstallation = getParent(self);
      return parent.metadataBucket;
    },

    get stateMachineArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get isGettingSchemas() {
      return self.loadingStage === LoadingStage.GettingSchemas && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get isMonitoringProgress() {
      return self.loadingStage === LoadingStage.MonitoringProgress && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get errorDetail() {
      return self._errorDetail;
    },

    get listObjects(): IObjectImportStatus[] {
      const result: readonly unknown[] = values(self.objects);
      return sortBy(result as IObjectImportStatus[], ["label", "name"]);
    },

    get objectsCount(): number {
      return self.objects.size;
    },

    get processPercentage(): number {
      const size = self.objects.size;
      let count = 0;

      self.objects.forEach((object) => {
        switch (object.status) {
          case ObjectImportStatusCode.Preparing:
            count += 0.25;
            break;
          case ObjectImportStatusCode.In_Progress:
            count += 0.5;
            break;
          case ObjectImportStatusCode.Importing:
            count += 0.75;
            break;
          case ObjectImportStatusCode.Successful:
            count += 1;
        }
      });

      if (size === 0) return 0; // Avoid division by zero
      return (count / size) * 100;
    },
  }))

  .views((self) => ({
    get isSuccess(): boolean {
      return isStoreReady(self) && !isStoreError(self) && isEmpty(self.errorDetail);
    },

    get totalRecords(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).rowCount, 0), 0);
    },

    get totalFields(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).columnCount, 0), 0);
    },

    get totalBytes(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).bytesWritten, 0), 0);
    },

    get totalTime(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).importSeconds, 0), 0);
    },
  }))

  .views((self) => ({
    get totalRecordsDisplay(): string {
      return niceNumber(self.totalRecords);
    },

    get totalFieldsDisplay(): string {
      return niceNumber(self.totalFields);
    },

    get totalBytesDisplay(): string {
      return niceByte(self.totalBytes);
    },

    get totalTimeDisplay(): string {
      return `${Math.round(self.totalTime)} sec`;
    },
  }))

  .actions((self) => ({
    setObject(rawObject: ObjectStatusInput) {
      if (self.objects.has(rawObject.name)) {
        self.objects.get(rawObject.name)?.setObjectStatus(rawObject);
      } else {
        self.objects.set(rawObject.name, rawObject);
      }

      return self.objects.get(rawObject.name);
    },

    setLoadingStage(stage: LoadingStage) {
      self.loadingStage = stage;
    },

    clearLoadingStage() {
      self.loadingStage = undefined;
    },

    setStartTime(time: number) {
      self.startTime = time;
    },
  }))

  .actions((self) => ({
    buildStateMachineUrl(): string {
      const region = self.region;
      const stateMachineArn = self.importWorkflowArn;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/statemachines/view/${encodeURIComponent(stateMachineArn)}`;
    },

    buildExecutionUrl(executionArn: string): string {
      const region = self.region;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/executions/details/${encodeURIComponent(executionArn)}`;
    },
  }))

  .actions((self) => ({
    async describeWorkflow(executionArn: string): Promise<ExecutionStatus> {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const info = await describeSFN({ accessKey, secretKey, region, executionArn });
        return info.status as ExecutionStatus;
      } catch (error) {
        self.runInAction(() => {
          // This is in markdown
          self._errorDetail = `We attempted to check the status of the execution of the state machine by calling the Step Functions DescribeExecution API.
          However, we got the following error:
### ${error}

### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}

### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})
* [Step Functions DescribeExecution API documentation](https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html)
          `;
        });
        throw error;
      }
    },

    async handleSFNFailure(executionArn: string, status: ExecutionStatus) {
      // In this method we display an informative error detail that provides enough context around the failure.
      // To do this, we will list all the execution history events and look for the overall execution event that had an error.
      // Since getting the history events is not critical, we will capture any errors from calling GetExecutionHistory
      // and not propagate them or have them affecting the overall status of the operation.

      let executionEvent: HistoryEvent;
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const events = await getSFNHistory({ accessKey, secretKey, region, executionArn });
        // We find the execution event that failed
        for (const event of events) {
          if (isExecutionFailureEvent(event)) {
            // We found the event
            executionEvent = event;
            break;
          }
        }
      } catch (error) {
        // Since getting the stack events is not critical, we will capture any errors from calling the describe stack events
        // call and not propagate them or have them affect the overall status of the operation.
        console.log(error);
      }

      self.runInAction(() => {
        // This is in markdown
        self._errorDetail = `The step function workflow failed with the following status:
### ${status}`;

        if (executionEvent) {
          const info = getExecutionFailureInfo(executionEvent);
          self._errorDetail = `${self._errorDetail}
### Execution Failure
* Cause: ${info.cause}
         `;
        }

        self._errorDetail = `${self._errorDetail}
### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}
        
### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})        
          `;
      });
      throw new Error(status);
    },

    async loadSchemas() {
      // We need to detect all the objects that we are expecting to import, this is done by looking at the schema files in S3
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as { name: string; label: string };
        self.setObject({ name: json.name, label: json.label ?? json.name });
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `schemas/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async pullProgress(executionArn: string) {
      // We pull that latest status of any object that has not reached the successful stage
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const runId = last(split(executionArn, ":"));
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        // An example of an S3 object key is
        // sf-metadata-u1puzajv0crm/state/runs/aaaaa-8d51-4dd3-bbf7-a81ee93cdc9a/Account.status.json
        // So, can split on "/" and take the last element which is "Account.status.json" and then
        // split on "." and take the first element which give us "Account"
        const name = first(split(last(split(s3Object.key, "/")), ".")) ?? ""; // Yes, we can regex but this is also okay, see explanation above
        try {
          const existing = self.objects.get(name);
          if (existing?.isSuccessful) return;

          const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as RemoteObject;
          const object = self.setObject({
            name,
            status: json.status,
            rowCount: json.metadata?.rowCount,
            columnCount: json.metadata?.columnCount,
            importSeconds: json.metadata?.importSeconds,
            bytesWritten: json.metadata?.bytesWritten,
          });

          object?.setStale(false);
        } catch (error) {
          // NOTE: it is important that we don't propagate this error, we don't want status reporting to halt deployment
          console.error(error);
        }
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `state/runs/${runId}/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async getLatestExecution() {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const stateMachineArn = self.stateMachineArn;

      return getLatestSFN({ accessKey, secretKey, region, stateMachineArn });
    },

    markStale() {
      self.objects.forEach((object) => {
        object.setStale(true);
      });
    },

    removeStale() {
      self.objects.forEach((object) => {
        if (object.stale) self.objects.delete(object.name); // It is safe to delete from a map while iterating
      });
    },
  }))

  .actions((self) => {
    const superAfterCreate = self.afterCreate;
    const superReset = self.reset;
    return {
      afterCreate() {
        superAfterCreate();
        self._errorDetail = "";
        self.loadingStage = undefined;
      },

      reset() {
        superReset();
        self._errorDetail = "";
        self.objects.clear();
        self.loadingStage = undefined;
      },

      async doLoad({ executionArn, startTime }: { executionArn?: string; startTime?: number } = {}) {
        self.loadingStage = LoadingStage.GettingSchemas;
        self._errorDetail = "";
        self.startTime = startTime ? startTime : self.startTime;
        self.markStale();

        if (!executionArn) {
          const latestExecution = await self.getLatestExecution();
          if (!latestExecution || !latestExecution.executionArn) {
            // If we don't have any, then there is nothing else to do
            self.setStartTime(0);
            self.clearLoadingStage();
            return;
          }
          self.setStartTime(latestExecution.startDate?.getTime() ?? 0);
          executionArn = latestExecution.executionArn;
        }

        // Load the schemas
        await self.loadSchemas();

        let status: ExecutionStatus;
        const shouldStop = (status: ExecutionStatus) => isFailureStatus(status) || isSuccessStatus(status);
        self.setLoadingStage(LoadingStage.MonitoringProgress);

        do {
          status = await self.describeWorkflow(executionArn);
          await delay(10);
          await self.pullProgress(executionArn);
        } while (!shouldStop(status));

        self.clearLoadingStage();
        if (isSuccessStatus(status)) {
          self.removeStale();
          return; // We are done with the success scenario
        }

        // We hit an SFN failure
        await self.handleSFNFailure(executionArn, status);
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IImportStatusStore extends Instance<typeof ImportStatusStore> {}

export function useImportStatusStore() {
  const installation = useInstallation();
  const store = installation.importStatusStore;

  useEffect(() => {
    // TODO - add a check for time and if it has been more than 10 minutes, then load the store
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return { isError: isStoreError(store), store };
}
