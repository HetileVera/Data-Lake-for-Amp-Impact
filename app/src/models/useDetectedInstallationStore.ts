import { isStoreError, isStoreLoading, isStoreReady, isStoreReLoading } from "models/BaseStore";

import { DetectedInstallationStore } from "./DetectedInstallationsStore";

// Create an instance of this store
const store = DetectedInstallationStore.create({});

export function useDetectedInstallationStore() {
  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
