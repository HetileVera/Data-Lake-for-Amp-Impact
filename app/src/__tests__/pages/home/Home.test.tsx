import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, RenderWithChakra } from "test-utils";

import { StoreProvider } from "AppContext";

import { Home } from "pages/home/Home";

import * as detectedInstallationStore from "models/useDetectedInstallationStore";
import { BrowserRouter, Router } from "react-router-dom";

describe("Home component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const tree = create(
      <RenderWithChakra>
        <StoreProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </StoreProvider>
      </RenderWithChakra>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
