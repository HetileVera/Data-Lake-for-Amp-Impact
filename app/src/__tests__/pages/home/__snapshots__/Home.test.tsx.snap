// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component snapshot 1`] = `
Array [
  <div
    className="css-1v55i00"
    data-testid="curved-box"
  />,
  <div
    className="chakra-container css-1szj4qk"
  >
    <div
      className="chakra-stack css-olw0lh"
    >
      <div
        className="chakra-stack css-18lyn6q"
      >
        <div
          className="css-1i3fafl"
        >
          <img
            className="chakra-image css-0"
            src="Salesforce_Corporate_Logo_RGB.png"
          />
        </div>
        <div
          className="chakra-stack__divider css-r2vkku"
        />
        <div
          className="css-nutmff"
        >
          <img
            className="chakra-image css-ir5tdx"
            src="AWS_logo_RGB_BLK.png"
          />
        </div>
      </div>
      <div
        className="chakra-stack css-1n7gsjr"
      >
        <div
          className="css-18n4g2t"
        >
          <h2
            className="chakra-heading css-1opsxio"
          >
            Datalake for Nonprofits, Powered by AWS
          </h2>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="chakra-container css-ms16h5"
  >
    <div
      className="css-57n2a7"
    >
      <div
        className="css-0"
      >
        <h2
          className="chakra-heading css-1lmbi92"
        >
          What to expect
        </h2>
      </div>
      <p
        className="chakra-text css-tvzngu"
      >
        In just a few steps, you can bring your Salesforce data to AWS and keep it in sync. Take a look at the summary of the steps needed, no worries though, we will guide you through each step.
      </p>
      <div
        className="css-1tt2enq"
      >
        <button
          className="chakra-button css-mwvyhu"
          onClick={[Function]}
          type="button"
        >
          Let's Go
          <span
            className="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden={true}
              className="chakra-icon css-onkibi"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              1
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Connect to your AWS account
          </div>
          <div
            className="css-1pip4vl"
          >
            We need your AWS admin credentials so that we can connect to your AWS account and provision the resources for the data lake. We will
  guide you through the process of obtaining the needed credentials
          </div>
        </div>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              2
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Connect to your Salesforce organization
          </div>
          <div
            className="css-1pip4vl"
          >
            For AWS to access your Salesforce data, we will need your authorization. We will guide you through the steps of setting up Amazon AppFlow which
    is a service that allows AWS to connect to Salesforce, you will be creating a Salesforce connection, signing in to Salesforce and authorizing access
          </div>
        </div>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              3
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Select data import options
          </div>
          <div
            className="css-1pip4vl"
          >
            In this step, you can pick the standard NPSP import options for your Salesforce data or customize these options to fit your needs
          </div>
        </div>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              4
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Review and confirm
          </div>
          <div
            className="css-1pip4vl"
          >
            You get a chance to review and confirm the settings you have provided
          </div>
        </div>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              5
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Sit back and relax
          </div>
          <div
            className="css-1pip4vl"
          >
            In this step, we will do all the necessary provisioning and data importing. You can sit back and relax and watch the progress of the data import
          </div>
        </div>
      </div>
      <div
        className="css-11ew3nf"
      >
        <div
          className="css-7628p4"
        >
          <div
            className="css-1g7wgva"
          >
            <span
              className="css-wr7jay"
            >
              6
            </span>
          </div>
          <div
            className="css-pfuu37"
          >
            Connect to Analytics
          </div>
          <div
            className="css-1pip4vl"
          >
            This is the final step, you can create users with credentials to access the data lake using Tableau
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-5f40t2"
    >
      <div
        className="css-0"
      >
        <h2
          className="chakra-heading css-5lipfx"
        >
          Resume
        </h2>
      </div>
      <p
        className="chakra-text css-tvzngu"
      >
        If you have already created a datalake using the EZ Data Lake app and would like to view or resume its setup, we got you covered. Click the button to reconnect to an existing datalake, we will ask you for your AWS admin credentials so that we can detect the existing datalake setup.
      </p>
      <div
        className="css-18022i5"
      >
        <button
          className="chakra-button css-1bex62d"
          onClick={[Function]}
          type="button"
        >
          Resume
          <span
            className="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden={true}
              fill="currentColor"
              focusable={false}
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M382.957 22.023c-2.618.033-5.138.96-7.26 3.082l-67.88 67.883 22.627 22.627 67.882-67.883c5.657-5.657 2.83-14.142-2.828-19.8-3.535-3.534-8.177-5.964-12.54-5.91zm-131.71 37.023l-21.92 21.92 203.648 203.647 21.92-21.92-11.315-11.315L262.56 70.36 251.25 59.047zm-23.335 45.96l-55.86 55.863v90.508l-22.628 22.63c-11.314 11.313-11.315 33.94.092 45.347l45.164 45.162c11.314 11.314 33.94 11.314 45.254 0l22.627-22.63h90.51l55.862-55.86-181.02-181.02zm245.555 7.527c-2.62.033-5.138.96-7.26 3.082l-67.88 67.883 22.626 22.626 67.883-67.882c5.657-5.657 2.83-14.142-2.828-19.8-3.536-3.535-8.177-5.965-12.54-5.91zM148.08 349.25l-128 128 12.727 12.728 128-128-12.727-12.727z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
  <div
    className="css-16kb41"
  >
    Â 
  </div>,
  <span />,
]
`;
